src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,208 @@
+        private readonly Action<SdamInformationEvent> _sdamInformationEventHandler;
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, IDnsResolver dnsResolver, string lookupDomainName, IEventSubscriber eventSubscriber, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _dnsResolver = Ensure.IsNotNull(dnsResolver, nameof(dnsResolver));
+            _lookupDomainName = EnsureLookupDomainNameIsValid(lookupDomainName);
+            _cancellationToken = cancellationToken;
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _service = "_mongodb._tcp." + _lookupDomainName;
+            _state = DnsMonitorState.Created;
+
+            eventSubscriber?.TryGetEventHandler(out _sdamInformationEventHandler);
+        }
+
+        // public properties
+        public DnsMonitorState State => _state;
+
+        public Exception UnhandledException => _unhandledException;
+
+        // public methods
+        public void Start()
+        {
+            _state = DnsMonitorState.Running;
+
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception exception)
+            {
+                _unhandledException = exception;
+
+                if (_sdamInformationEventHandler != null)
+                {
+                    var message = $"Unhandled exception in DnsMonitor: {exception}.";
+                    var sdamInformationEvent = new SdamInformationEvent(() => message);
+                    _sdamInformationEventHandler(sdamInformationEvent);
+                }
+
+                _state = DnsMonitorState.Failed;
+                return;
+            }
+
+            _state = DnsMonitorState.Stopped;
+        }
+
+        // private methods
+        private TimeSpan ComputeRescanDelay(List<SrvRecord> srvRecords)
+        {
+            var delay = TimeSpan.FromSeconds(60);
+
+            if (srvRecords.Count > 0)
+            {
+                var minTimeToLive = srvRecords.Select(s => s.TimeToLive).Min();
+                if (minTimeToLive > delay)
+                {
+                    delay = minTimeToLive;
+                }
+            }
+
+            return delay;
+        }
+
+        private List<DnsEndPoint> GetValidEndPoints(List<SrvRecord> srvRecords)
+        {
+            var validEndPoints = new List<DnsEndPoint>();
+
+            foreach (var srvRecord in srvRecords)
+            {
+                var endPoint = srvRecord.EndPoint;
+                var host = endPoint.Host;
+                if (host.EndsWith("."))
+                {
+                    host = host.Substring(0, host.Length - 1);
+                    endPoint = new DnsEndPoint(host, endPoint.Port);
+                }
+
+                if (IsValidHost(host))
```

Comment:
It seems like this logic should be the same as what's in ConnectionString.ValidateResolvedHosts.  Am I wrong about that?

Reply:
Good question. The implementation of IsValidHost in this file is one line long, and the implementation of ValidateResolvedHosts in ConnectionString is 43 lines long. That's a surprisingly large difference for two methods that presumably do the same thing.

I thought all we had to validate was that the resolved hostnamed shared the same parent domain from the lookup host name.

Reply:
From the spec:

    A driver MUST verify that the host names returned through SRV records have the same parent {domainname}.

I have no idea why the implementation of ConnectionString.ValidateResolvedHosts is so complicated. The spec only requires that the host have the same parent domain, which is a simple string comparison (the actual implementation is not here, but in the IsValid method below).

Reply:
I think it's to catch situations like this:

     var parentDomain = "parent.com"
     var resolvedDomain = "badparent.com"

I think there are tests in the original spec that would catch that (parent-part-mismatch-*.json) but those tests don't execute on this code, only the code in ConnectionString.ValidateResolvedHosts.

Can we figure out a way to share this logic, or at worst copy it?